pkg/status/combinedstatus-resolution.go

```diff
@@ -0,0 +1,308 @@
+//   - When a workstatus is added to / modified in the combinedstatus
+//     resolution, every clause in every statuscollector is evaluated against
+//     the workstatus. The caching of collector-clause -> evaluation makes
+//     calculating combinedFields efficient if present.
+//
+// - Ownership:
+//   - A combinedstatus resolution is owned by the binding that it is
+//     associated with.
+type combinedStatusResolution struct {
+	sync.RWMutex
+
+	// statusCollectorNameToData is a map of status collector names to
+	// their corresponding data.
+	// If a status collector name is mapped to nil, it means that the status
+	// collector has not been processed.
+	statusCollectorNameToData map[string]*statusCollectorData
+	// workStatusToStatusCollectorName is a map of workstatus references to
+	// a set of statuscollector names that they contribute to.
+	workStatusToStatusCollectorName map[util.ObjectIdentifier]sets.Set[string]
+
+	// ownerReference identifies the binding that this resolution is
+	// associated with as an owning object.
+	ownerReference *metav1.OwnerReference
+}
+
+// statusCollectorData is a struct that represents the data of a status
+// collector in a combinedstatus resolution.
+type statusCollectorData struct {
+	*statusCollectorClauses
+
+	// workStatusRefToData is a map of workstatus references to their
+	// corresponding work status data.
+	// A workstatus reference is identical to its source object's identifier,
+	// apart from the namespace, which is set to the associated clusterName.
+	// If a workstatus reference is mapped to nil, it means that the workstatus
+	// has not been evaluated against the statuscollector's clauses.
+	workStatusRefToData map[util.ObjectIdentifier]workStatusData
+}
+
+// workStatusData is a struct that represents the evaluation of a workstatus
+// against a statuscollector's clauses.
+type workStatusData struct {
+	// filterEval is a boolean that represents the evaluation of the filter
+	// expression against the workstatus. If true, the workstatus is included
+	// in the statuscollector's result.
+	filterEval bool
+
+	// groupByEval is a map of groupBy expression names to their evaluated values.
+	groupByEval map[string]any
+	// combinedFieldsEval is a map of combinedFields expression names to their
+	// evaluated values. CombinedField types:
+	// - COUNT: the number of workstatuses per groupBy value. If groupBy is
+	//   empty, the count is the number of workstatuses.
+	// - SUM: the sum of the values of the workstatuses groupBy values.
+	// - AVG: the average of the values of the workstatuses groupBy values.
+	// - MIN: the minimum value of the workstatuses groupBy values.
+	// - MAX: the maximum value of the workstatuses groupBy values.
+	combinedFieldsEval map[string]any
+
+	selectEval map[string]any
+}
+
+// setStatusCollectors sets ALL the statuscollectors relevant to the
+// combinedstatus resolution.
+// The given map is expected to be thread-safe, its values contain valid
```

Comment:
A primitive map is not thread-safe. I think you mean to say that nothing mutates the map during a call on this method.